So the fail was false already before my changes, I just removed it.
Output was this :
https://travis-ci.org/BakedSoftware/go-parameters/builds/161323505

```
--- FAIL: TestParseJSONBodyMux (0.00s)
    params_test.go:238: Key: 'id' not found
FAIL
```

--- a/http_handlers.go
+++ b/http_handlers.go
@@ -99,8 +99,8 @@ func EnableGZIP(fn httprouter.Handle) httprouter.Handle {
 		}
 		w.Header().Set("Content-Encoding", "gzip")
 		gz := gzip.NewWriter(w)
-		defer gz.Close()
 		gzr := gzipResponseWriter{Writer: gz, ResponseWriter: w}
 		fn(gzr, r, p)
+		gz.Close()
 	}
 }

--- a/params.go
+++ b/params.go
@@ -476,7 +476,6 @@ func MakeParsedReq(fn http.HandlerFunc) http.HandlerFunc {
 
 func MakeHTTPRouterParsedReq(fn httprouter.Handle) httprouter.Handle {
 	return func(rw http.ResponseWriter, r *http.Request, p httprouter.Params) {
-		log.Print("MakeHTTPRouterParsedReq ", r.Header.Get("X-Request-ID"))
 		r = r.WithContext(context.WithValue(r.Context(), paramsKey, ParseParams(r)))
 		params := GetParams(r)
 		for _, param := range p {
@@ -492,7 +491,6 @@ func MakeHTTPRouterParsedReq(fn httprouter.Handle) httprouter.Handle {
 				params.Values[param.Key] = param.Value
 			}
 		}
-		log.Print("MakeHTTPRouterParsedReq.fn ! ", r.Header.Get("X-Request-ID"))
 		fn(rw, r, p)
 	}
 }

--- a/params.go
+++ b/params.go
@@ -9,6 +9,7 @@ package parameters
 
 import (
 	"bytes"
+	"context"
 	"encoding/base64"
 	"encoding/json"
 	"io"
@@ -467,16 +468,17 @@ func (p *Params) GetJSON(key string) map[string]interface{} {
 }
 
 func MakeParsedReq(fn http.HandlerFunc) http.HandlerFunc {
-	return func(rw http.ResponseWriter, req *http.Request) {
-		ParseParams(req)
-		fn(rw, req)
+	return func(rw http.ResponseWriter, r *http.Request) {
+		r = r.WithContext(context.WithValue(r.Context(), paramsKey, ParseParams(r)))
+		fn(rw, r)
 	}
 }
 
 func MakeHTTPRouterParsedReq(fn httprouter.Handle) httprouter.Handle {
-	return func(rw http.ResponseWriter, req *http.Request, p httprouter.Params) {
-		ParseParams(req)
-		params := GetParams(req)
+	return func(rw http.ResponseWriter, r *http.Request, p httprouter.Params) {
+		log.Print("MakeHTTPRouterParsedReq ", r.Header.Get("X-Request-ID"))
+		r = r.WithContext(context.WithValue(r.Context(), paramsKey, ParseParams(r)))
+		params := GetParams(r)
 		for _, param := range p {
 			const ID = "id"
 			if strings.Contains(param.Key, ID) {
@@ -490,7 +492,7 @@ func MakeHTTPRouterParsedReq(fn httprouter.Handle) httprouter.Handle {
 				params.Values[param.Key] = param.Value
 			}
 		}
-		fn(rw, req, p)
+		fn(rw, r, p)
 	}
 }
 

